{"nbformat":4,"nbformat_minor":0,"metadata":{"kernelspec":{"display_name":"R","language":"R","name":"ir"},"language_info":{"codemirror_mode":"r","file_extension":".r","mimetype":"text/x-r-source","name":"R","pygments_lexer":"r","version":"3.4.1"},"colab":{"name":"D1-1 Introduction to Bioconductor.ipynb","provenance":[{"file_id":"15iemQJ3R1OXzmCNU8352zk0vy35P9DAI","timestamp":1603749727334}],"collapsed_sections":[],"toc_visible":true}},"cells":[{"cell_type":"markdown","metadata":{"id":"oh5M0jvVyC97"},"source":["## 1. Introduction to Bioconductor in R\n"]},{"cell_type":"code","metadata":{"dc":{"key":"3"},"tags":["sample_code"],"id":"0koPxrfX1Ipb","executionInfo":{"status":"ok","timestamp":1603763871430,"user_tz":420,"elapsed":4668,"user":{"displayName":"Jun Meng","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gjs1DreGo6bVjgzln2TQWojZ2ZE86B5qF7-nsNg8nw=s64","userId":"16895417534398299616"}},"outputId":"a3416005-7190-4e2e-db8f-c8055579c68c","colab":{"base_uri":"https://localhost:8080/","height":102}},"source":["if (!requireNamespace(\"BioManager\"))\n","  install.packages(\"BiocManager\")"],"execution_count":null,"outputs":[{"output_type":"stream","text":["Loading required namespace: BioManager\n","\n","Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n"],"name":"stderr"}]},{"cell_type":"markdown","metadata":{"id":"4UBrkDHa9sVg"},"source":["### 1.1 Install Packages via BiocManager"]},{"cell_type":"code","metadata":{"id":"Kyl6qq968xop","outputId":"821a651a-4524-4d05-ee22-846540ef378e","colab":{"base_uri":"https://localhost:8080/","height":258}},"source":["BiocManager::install(\"BSgenome\")"],"execution_count":null,"outputs":[{"output_type":"stream","text":["Bioconductor version 3.11 (BiocManager 1.30.10), R 4.0.3 (2020-10-10)\n","\n","Installing package(s) 'BiocVersion', 'BSgenome'\n","\n","also installing the dependencies ‘formatR’, ‘Biobase’, ‘DelayedArray’, ‘lambda.r’, ‘futile.options’, ‘SummarizedExperiment’, ‘futile.logger’, ‘snow’, ‘RCurl’, ‘GenomeInfoDbData’, ‘XML’, ‘zlibbioc’, ‘GenomicAlignments’, ‘bitops’, ‘BiocParallel’, ‘Rhtslib’, ‘BiocGenerics’, ‘S4Vectors’, ‘IRanges’, ‘GenomeInfoDb’, ‘GenomicRanges’, ‘Biostrings’, ‘rtracklayer’, ‘matrixStats’, ‘XVector’, ‘Rsamtools’\n","\n","\n","Warning message in install.packages(...):\n","“installation of package ‘matrixStats’ had non-zero exit status”\n","Warning message in install.packages(...):\n","“installation of package ‘DelayedArray’ had non-zero exit status”\n","Warning message in install.packages(...):\n","“installation of package ‘SummarizedExperiment’ had non-zero exit status”\n"],"name":"stderr"}]},{"cell_type":"code","metadata":{"id":"7TevWDQn8x1T"},"source":["library(\"BSgenome\")\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"JpgSQ-Tk8x33","executionInfo":{"status":"ok","timestamp":1603753370808,"user_tz":420,"elapsed":305,"user":{"displayName":"Jun Meng","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gjs1DreGo6bVjgzln2TQWojZ2ZE86B5qF7-nsNg8nw=s64","userId":"16895417534398299616"}},"outputId":"7a38bb0f-37ac-4a70-b403-7fd0f3ceb060","colab":{"base_uri":"https://localhost:8080/","height":34}},"source":["# Check the version of the BSgenome package\n","packageVersion(\"BSgenome\")"],"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["[1] ‘1.56.0’"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"yhQIR7xq8x-W","executionInfo":{"status":"ok","timestamp":1603753410754,"user_tz":420,"elapsed":382,"user":{"displayName":"Jun Meng","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gjs1DreGo6bVjgzln2TQWojZ2ZE86B5qF7-nsNg8nw=s64","userId":"16895417534398299616"}},"outputId":"85913c9b-3b3c-4213-df3d-42793c6c403b","colab":{"base_uri":"https://localhost:8080/","height":442}},"source":["# check the formal definition of this class\n","\n","showClass(\"BSgenome\")"],"execution_count":null,"outputs":[{"output_type":"stream","text":["Class \"BSgenome\" [package \"BSgenome\"]\n","\n","Slots:\n","                                                                     \n","Name:               pkgname     single_sequences   multiple_sequences\n","Class:            character OnDiskNamedSequences        RdaCollection\n","                                                                     \n","Name:            source_url        user_seqnames   injectSNPs_handler\n","Class:            character            character    InjectSNPsHandler\n","                                                                     \n","Name:           .seqs_cache         .link_counts        nmask_per_seq\n","Class:          environment          environment              integer\n","                                                                     \n","Name:                 masks             organism          common_name\n","Class:        RdaCollection            character            character\n","                                                                     \n","Name:              provider     provider_version         release_date\n","Class:            character            character            character\n","                                                \n","Name:          release_name              seqinfo\n","Class:            character              Seqinfo\n","\n","Extends: \"GenomeDescription\"\n","\n","Known Subclasses: \"MaskedBSgenome\"\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"ieeQGzmTJ4iH"},"source":["### 1.2 Interaction with classes"]},{"cell_type":"code","metadata":{"id":"hoaOMxeq8yFv","executionInfo":{"status":"ok","timestamp":1603753995945,"user_tz":420,"elapsed":1218,"user":{"displayName":"Jun Meng","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gjs1DreGo6bVjgzln2TQWojZ2ZE86B5qF7-nsNg8nw=s64","userId":"16895417534398299616"}},"outputId":"4edb9a53-8f50-4205-8c52-b9b4fab665fb","colab":{"base_uri":"https://localhost:8080/","height":190}},"source":["install.packages(\" BSgenome.Scerevisiae.UCSC.sacCer3\")"],"execution_count":null,"outputs":[{"output_type":"stream","text":["Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n","Warning message:\n","“package ‘ BSgenome.Scerevisiae.UCSC.sacCer3’ is not available for this version of R\n","\n","A version of this package for your version of R might be available elsewhere,\n","see the ideas at\n","https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages”\n"],"name":"stderr"}]},{"cell_type":"code","metadata":{"id":"Q3RohY0F8x7N","executionInfo":{"status":"ok","timestamp":1603754192409,"user_tz":420,"elapsed":674,"user":{"displayName":"Jun Meng","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gjs1DreGo6bVjgzln2TQWojZ2ZE86B5qF7-nsNg8nw=s64","userId":"16895417534398299616"}},"outputId":"ff988ac3-ab0b-4d9d-c484-27fb3407aed1","colab":{"base_uri":"https://localhost:8080/","height":850}},"source":["available.genomes()"],"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":[" [1] \"BSgenome.Alyrata.JGI.v1\"                  \n"," [2] \"BSgenome.Amellifera.BeeBase.assembly4\"    \n"," [3] \"BSgenome.Amellifera.UCSC.apiMel2\"         \n"," [4] \"BSgenome.Amellifera.UCSC.apiMel2.masked\"  \n"," [5] \"BSgenome.Aofficinalis.NCBI.V1\"            \n"," [6] \"BSgenome.Athaliana.TAIR.04232008\"         \n"," [7] \"BSgenome.Athaliana.TAIR.TAIR9\"            \n"," [8] \"BSgenome.Btaurus.UCSC.bosTau3\"            \n"," [9] \"BSgenome.Btaurus.UCSC.bosTau3.masked\"     \n","[10] \"BSgenome.Btaurus.UCSC.bosTau4\"            \n","[11] \"BSgenome.Btaurus.UCSC.bosTau4.masked\"     \n","[12] \"BSgenome.Btaurus.UCSC.bosTau6\"            \n","[13] \"BSgenome.Btaurus.UCSC.bosTau6.masked\"     \n","[14] \"BSgenome.Btaurus.UCSC.bosTau8\"            \n","[15] \"BSgenome.Btaurus.UCSC.bosTau9\"            \n","[16] \"BSgenome.Carietinum.NCBI.v1\"              \n","[17] \"BSgenome.Celegans.UCSC.ce10\"              \n","[18] \"BSgenome.Celegans.UCSC.ce11\"              \n","[19] \"BSgenome.Celegans.UCSC.ce2\"               \n","[20] \"BSgenome.Celegans.UCSC.ce6\"               \n","[21] \"BSgenome.Cfamiliaris.UCSC.canFam2\"        \n","[22] \"BSgenome.Cfamiliaris.UCSC.canFam2.masked\" \n","[23] \"BSgenome.Cfamiliaris.UCSC.canFam3\"        \n","[24] \"BSgenome.Cfamiliaris.UCSC.canFam3.masked\" \n","[25] \"BSgenome.Cjacchus.UCSC.calJac3\"           \n","[26] \"BSgenome.Dmelanogaster.UCSC.dm2\"          \n","[27] \"BSgenome.Dmelanogaster.UCSC.dm2.masked\"   \n","[28] \"BSgenome.Dmelanogaster.UCSC.dm3\"          \n","[29] \"BSgenome.Dmelanogaster.UCSC.dm3.masked\"   \n","[30] \"BSgenome.Dmelanogaster.UCSC.dm6\"          \n","[31] \"BSgenome.Drerio.UCSC.danRer10\"            \n","[32] \"BSgenome.Drerio.UCSC.danRer11\"            \n","[33] \"BSgenome.Drerio.UCSC.danRer5\"             \n","[34] \"BSgenome.Drerio.UCSC.danRer5.masked\"      \n","[35] \"BSgenome.Drerio.UCSC.danRer6\"             \n","[36] \"BSgenome.Drerio.UCSC.danRer6.masked\"      \n","[37] \"BSgenome.Drerio.UCSC.danRer7\"             \n","[38] \"BSgenome.Drerio.UCSC.danRer7.masked\"      \n","[39] \"BSgenome.Dvirilis.Ensembl.dvircaf1\"       \n","[40] \"BSgenome.Ecoli.NCBI.20080805\"             \n","[41] \"BSgenome.Gaculeatus.UCSC.gasAcu1\"         \n","[42] \"BSgenome.Gaculeatus.UCSC.gasAcu1.masked\"  \n","[43] \"BSgenome.Ggallus.UCSC.galGal3\"            \n","[44] \"BSgenome.Ggallus.UCSC.galGal3.masked\"     \n","[45] \"BSgenome.Ggallus.UCSC.galGal4\"            \n","[46] \"BSgenome.Ggallus.UCSC.galGal4.masked\"     \n","[47] \"BSgenome.Ggallus.UCSC.galGal5\"            \n","[48] \"BSgenome.Ggallus.UCSC.galGal6\"            \n","[49] \"BSgenome.Hsapiens.1000genomes.hs37d5\"     \n","[50] \"BSgenome.Hsapiens.NCBI.GRCh38\"            \n","[51] \"BSgenome.Hsapiens.UCSC.hg17\"              \n","[52] \"BSgenome.Hsapiens.UCSC.hg17.masked\"       \n","[53] \"BSgenome.Hsapiens.UCSC.hg18\"              \n","[54] \"BSgenome.Hsapiens.UCSC.hg18.masked\"       \n","[55] \"BSgenome.Hsapiens.UCSC.hg19\"              \n","[56] \"BSgenome.Hsapiens.UCSC.hg19.masked\"       \n","[57] \"BSgenome.Hsapiens.UCSC.hg38\"              \n","[58] \"BSgenome.Hsapiens.UCSC.hg38.masked\"       \n","[59] \"BSgenome.Mdomestica.UCSC.monDom5\"         \n","[60] \"BSgenome.Mfascicularis.NCBI.5.0\"          \n","[61] \"BSgenome.Mfuro.UCSC.musFur1\"              \n","[62] \"BSgenome.Mmulatta.UCSC.rheMac10\"          \n","[63] \"BSgenome.Mmulatta.UCSC.rheMac2\"           \n","[64] \"BSgenome.Mmulatta.UCSC.rheMac2.masked\"    \n","[65] \"BSgenome.Mmulatta.UCSC.rheMac3\"           \n","[66] \"BSgenome.Mmulatta.UCSC.rheMac3.masked\"    \n","[67] \"BSgenome.Mmulatta.UCSC.rheMac8\"           \n","[68] \"BSgenome.Mmusculus.UCSC.mm10\"             \n","[69] \"BSgenome.Mmusculus.UCSC.mm10.masked\"      \n","[70] \"BSgenome.Mmusculus.UCSC.mm8\"              \n","[71] \"BSgenome.Mmusculus.UCSC.mm8.masked\"       \n","[72] \"BSgenome.Mmusculus.UCSC.mm9\"              \n","[73] \"BSgenome.Mmusculus.UCSC.mm9.masked\"       \n","[74] \"BSgenome.Osativa.MSU.MSU7\"                \n","[75] \"BSgenome.Ptroglodytes.UCSC.panTro2\"       \n","[76] \"BSgenome.Ptroglodytes.UCSC.panTro2.masked\"\n","[77] \"BSgenome.Ptroglodytes.UCSC.panTro3\"       \n","[78] \"BSgenome.Ptroglodytes.UCSC.panTro3.masked\"\n","[79] \"BSgenome.Ptroglodytes.UCSC.panTro5\"       \n","[80] \"BSgenome.Ptroglodytes.UCSC.panTro6\"       \n","[81] \"BSgenome.Rnorvegicus.UCSC.rn4\"            \n","[82] \"BSgenome.Rnorvegicus.UCSC.rn4.masked\"     \n","[83] \"BSgenome.Rnorvegicus.UCSC.rn5\"            \n","[84] \"BSgenome.Rnorvegicus.UCSC.rn5.masked\"     \n","[85] \"BSgenome.Rnorvegicus.UCSC.rn6\"            \n","[86] \"BSgenome.Scerevisiae.UCSC.sacCer1\"        \n","[87] \"BSgenome.Scerevisiae.UCSC.sacCer2\"        \n","[88] \"BSgenome.Scerevisiae.UCSC.sacCer3\"        \n","[89] \"BSgenome.Sscrofa.UCSC.susScr11\"           \n","[90] \"BSgenome.Sscrofa.UCSC.susScr3\"            \n","[91] \"BSgenome.Sscrofa.UCSC.susScr3.masked\"     \n","[92] \"BSgenome.Tgondii.ToxoDB.7.0\"              \n","[93] \"BSgenome.Tguttata.UCSC.taeGut1\"           \n","[94] \"BSgenome.Tguttata.UCSC.taeGut1.masked\"    \n","[95] \"BSgenome.Tguttata.UCSC.taeGut2\"           \n","[96] \"BSgenome.Vvinifera.URGI.IGGP12Xv0\"        \n","[97] \"BSgenome.Vvinifera.URGI.IGGP12Xv2\"        \n","[98] \"BSgenome.Vvinifera.URGI.IGGP8X\"           "],"text/latex":"\\begin{enumerate*}\n\\item 'BSgenome.Alyrata.JGI.v1'\n\\item 'BSgenome.Amellifera.BeeBase.assembly4'\n\\item 'BSgenome.Amellifera.UCSC.apiMel2'\n\\item 'BSgenome.Amellifera.UCSC.apiMel2.masked'\n\\item 'BSgenome.Aofficinalis.NCBI.V1'\n\\item 'BSgenome.Athaliana.TAIR.04232008'\n\\item 'BSgenome.Athaliana.TAIR.TAIR9'\n\\item 'BSgenome.Btaurus.UCSC.bosTau3'\n\\item 'BSgenome.Btaurus.UCSC.bosTau3.masked'\n\\item 'BSgenome.Btaurus.UCSC.bosTau4'\n\\item 'BSgenome.Btaurus.UCSC.bosTau4.masked'\n\\item 'BSgenome.Btaurus.UCSC.bosTau6'\n\\item 'BSgenome.Btaurus.UCSC.bosTau6.masked'\n\\item 'BSgenome.Btaurus.UCSC.bosTau8'\n\\item 'BSgenome.Btaurus.UCSC.bosTau9'\n\\item 'BSgenome.Carietinum.NCBI.v1'\n\\item 'BSgenome.Celegans.UCSC.ce10'\n\\item 'BSgenome.Celegans.UCSC.ce11'\n\\item 'BSgenome.Celegans.UCSC.ce2'\n\\item 'BSgenome.Celegans.UCSC.ce6'\n\\item 'BSgenome.Cfamiliaris.UCSC.canFam2'\n\\item 'BSgenome.Cfamiliaris.UCSC.canFam2.masked'\n\\item 'BSgenome.Cfamiliaris.UCSC.canFam3'\n\\item 'BSgenome.Cfamiliaris.UCSC.canFam3.masked'\n\\item 'BSgenome.Cjacchus.UCSC.calJac3'\n\\item 'BSgenome.Dmelanogaster.UCSC.dm2'\n\\item 'BSgenome.Dmelanogaster.UCSC.dm2.masked'\n\\item 'BSgenome.Dmelanogaster.UCSC.dm3'\n\\item 'BSgenome.Dmelanogaster.UCSC.dm3.masked'\n\\item 'BSgenome.Dmelanogaster.UCSC.dm6'\n\\item 'BSgenome.Drerio.UCSC.danRer10'\n\\item 'BSgenome.Drerio.UCSC.danRer11'\n\\item 'BSgenome.Drerio.UCSC.danRer5'\n\\item 'BSgenome.Drerio.UCSC.danRer5.masked'\n\\item 'BSgenome.Drerio.UCSC.danRer6'\n\\item 'BSgenome.Drerio.UCSC.danRer6.masked'\n\\item 'BSgenome.Drerio.UCSC.danRer7'\n\\item 'BSgenome.Drerio.UCSC.danRer7.masked'\n\\item 'BSgenome.Dvirilis.Ensembl.dvircaf1'\n\\item 'BSgenome.Ecoli.NCBI.20080805'\n\\item 'BSgenome.Gaculeatus.UCSC.gasAcu1'\n\\item 'BSgenome.Gaculeatus.UCSC.gasAcu1.masked'\n\\item 'BSgenome.Ggallus.UCSC.galGal3'\n\\item 'BSgenome.Ggallus.UCSC.galGal3.masked'\n\\item 'BSgenome.Ggallus.UCSC.galGal4'\n\\item 'BSgenome.Ggallus.UCSC.galGal4.masked'\n\\item 'BSgenome.Ggallus.UCSC.galGal5'\n\\item 'BSgenome.Ggallus.UCSC.galGal6'\n\\item 'BSgenome.Hsapiens.1000genomes.hs37d5'\n\\item 'BSgenome.Hsapiens.NCBI.GRCh38'\n\\item 'BSgenome.Hsapiens.UCSC.hg17'\n\\item 'BSgenome.Hsapiens.UCSC.hg17.masked'\n\\item 'BSgenome.Hsapiens.UCSC.hg18'\n\\item 'BSgenome.Hsapiens.UCSC.hg18.masked'\n\\item 'BSgenome.Hsapiens.UCSC.hg19'\n\\item 'BSgenome.Hsapiens.UCSC.hg19.masked'\n\\item 'BSgenome.Hsapiens.UCSC.hg38'\n\\item 'BSgenome.Hsapiens.UCSC.hg38.masked'\n\\item 'BSgenome.Mdomestica.UCSC.monDom5'\n\\item 'BSgenome.Mfascicularis.NCBI.5.0'\n\\item 'BSgenome.Mfuro.UCSC.musFur1'\n\\item 'BSgenome.Mmulatta.UCSC.rheMac10'\n\\item 'BSgenome.Mmulatta.UCSC.rheMac2'\n\\item 'BSgenome.Mmulatta.UCSC.rheMac2.masked'\n\\item 'BSgenome.Mmulatta.UCSC.rheMac3'\n\\item 'BSgenome.Mmulatta.UCSC.rheMac3.masked'\n\\item 'BSgenome.Mmulatta.UCSC.rheMac8'\n\\item 'BSgenome.Mmusculus.UCSC.mm10'\n\\item 'BSgenome.Mmusculus.UCSC.mm10.masked'\n\\item 'BSgenome.Mmusculus.UCSC.mm8'\n\\item 'BSgenome.Mmusculus.UCSC.mm8.masked'\n\\item 'BSgenome.Mmusculus.UCSC.mm9'\n\\item 'BSgenome.Mmusculus.UCSC.mm9.masked'\n\\item 'BSgenome.Osativa.MSU.MSU7'\n\\item 'BSgenome.Ptroglodytes.UCSC.panTro2'\n\\item 'BSgenome.Ptroglodytes.UCSC.panTro2.masked'\n\\item 'BSgenome.Ptroglodytes.UCSC.panTro3'\n\\item 'BSgenome.Ptroglodytes.UCSC.panTro3.masked'\n\\item 'BSgenome.Ptroglodytes.UCSC.panTro5'\n\\item 'BSgenome.Ptroglodytes.UCSC.panTro6'\n\\item 'BSgenome.Rnorvegicus.UCSC.rn4'\n\\item 'BSgenome.Rnorvegicus.UCSC.rn4.masked'\n\\item 'BSgenome.Rnorvegicus.UCSC.rn5'\n\\item 'BSgenome.Rnorvegicus.UCSC.rn5.masked'\n\\item 'BSgenome.Rnorvegicus.UCSC.rn6'\n\\item 'BSgenome.Scerevisiae.UCSC.sacCer1'\n\\item 'BSgenome.Scerevisiae.UCSC.sacCer2'\n\\item 'BSgenome.Scerevisiae.UCSC.sacCer3'\n\\item 'BSgenome.Sscrofa.UCSC.susScr11'\n\\item 'BSgenome.Sscrofa.UCSC.susScr3'\n\\item 'BSgenome.Sscrofa.UCSC.susScr3.masked'\n\\item 'BSgenome.Tgondii.ToxoDB.7.0'\n\\item 'BSgenome.Tguttata.UCSC.taeGut1'\n\\item 'BSgenome.Tguttata.UCSC.taeGut1.masked'\n\\item 'BSgenome.Tguttata.UCSC.taeGut2'\n\\item 'BSgenome.Vvinifera.URGI.IGGP12Xv0'\n\\item 'BSgenome.Vvinifera.URGI.IGGP12Xv2'\n\\item 'BSgenome.Vvinifera.URGI.IGGP8X'\n\\end{enumerate*}\n","text/markdown":"1. 'BSgenome.Alyrata.JGI.v1'\n2. 'BSgenome.Amellifera.BeeBase.assembly4'\n3. 'BSgenome.Amellifera.UCSC.apiMel2'\n4. 'BSgenome.Amellifera.UCSC.apiMel2.masked'\n5. 'BSgenome.Aofficinalis.NCBI.V1'\n6. 'BSgenome.Athaliana.TAIR.04232008'\n7. 'BSgenome.Athaliana.TAIR.TAIR9'\n8. 'BSgenome.Btaurus.UCSC.bosTau3'\n9. 'BSgenome.Btaurus.UCSC.bosTau3.masked'\n10. 'BSgenome.Btaurus.UCSC.bosTau4'\n11. 'BSgenome.Btaurus.UCSC.bosTau4.masked'\n12. 'BSgenome.Btaurus.UCSC.bosTau6'\n13. 'BSgenome.Btaurus.UCSC.bosTau6.masked'\n14. 'BSgenome.Btaurus.UCSC.bosTau8'\n15. 'BSgenome.Btaurus.UCSC.bosTau9'\n16. 'BSgenome.Carietinum.NCBI.v1'\n17. 'BSgenome.Celegans.UCSC.ce10'\n18. 'BSgenome.Celegans.UCSC.ce11'\n19. 'BSgenome.Celegans.UCSC.ce2'\n20. 'BSgenome.Celegans.UCSC.ce6'\n21. 'BSgenome.Cfamiliaris.UCSC.canFam2'\n22. 'BSgenome.Cfamiliaris.UCSC.canFam2.masked'\n23. 'BSgenome.Cfamiliaris.UCSC.canFam3'\n24. 'BSgenome.Cfamiliaris.UCSC.canFam3.masked'\n25. 'BSgenome.Cjacchus.UCSC.calJac3'\n26. 'BSgenome.Dmelanogaster.UCSC.dm2'\n27. 'BSgenome.Dmelanogaster.UCSC.dm2.masked'\n28. 'BSgenome.Dmelanogaster.UCSC.dm3'\n29. 'BSgenome.Dmelanogaster.UCSC.dm3.masked'\n30. 'BSgenome.Dmelanogaster.UCSC.dm6'\n31. 'BSgenome.Drerio.UCSC.danRer10'\n32. 'BSgenome.Drerio.UCSC.danRer11'\n33. 'BSgenome.Drerio.UCSC.danRer5'\n34. 'BSgenome.Drerio.UCSC.danRer5.masked'\n35. 'BSgenome.Drerio.UCSC.danRer6'\n36. 'BSgenome.Drerio.UCSC.danRer6.masked'\n37. 'BSgenome.Drerio.UCSC.danRer7'\n38. 'BSgenome.Drerio.UCSC.danRer7.masked'\n39. 'BSgenome.Dvirilis.Ensembl.dvircaf1'\n40. 'BSgenome.Ecoli.NCBI.20080805'\n41. 'BSgenome.Gaculeatus.UCSC.gasAcu1'\n42. 'BSgenome.Gaculeatus.UCSC.gasAcu1.masked'\n43. 'BSgenome.Ggallus.UCSC.galGal3'\n44. 'BSgenome.Ggallus.UCSC.galGal3.masked'\n45. 'BSgenome.Ggallus.UCSC.galGal4'\n46. 'BSgenome.Ggallus.UCSC.galGal4.masked'\n47. 'BSgenome.Ggallus.UCSC.galGal5'\n48. 'BSgenome.Ggallus.UCSC.galGal6'\n49. 'BSgenome.Hsapiens.1000genomes.hs37d5'\n50. 'BSgenome.Hsapiens.NCBI.GRCh38'\n51. 'BSgenome.Hsapiens.UCSC.hg17'\n52. 'BSgenome.Hsapiens.UCSC.hg17.masked'\n53. 'BSgenome.Hsapiens.UCSC.hg18'\n54. 'BSgenome.Hsapiens.UCSC.hg18.masked'\n55. 'BSgenome.Hsapiens.UCSC.hg19'\n56. 'BSgenome.Hsapiens.UCSC.hg19.masked'\n57. 'BSgenome.Hsapiens.UCSC.hg38'\n58. 'BSgenome.Hsapiens.UCSC.hg38.masked'\n59. 'BSgenome.Mdomestica.UCSC.monDom5'\n60. 'BSgenome.Mfascicularis.NCBI.5.0'\n61. 'BSgenome.Mfuro.UCSC.musFur1'\n62. 'BSgenome.Mmulatta.UCSC.rheMac10'\n63. 'BSgenome.Mmulatta.UCSC.rheMac2'\n64. 'BSgenome.Mmulatta.UCSC.rheMac2.masked'\n65. 'BSgenome.Mmulatta.UCSC.rheMac3'\n66. 'BSgenome.Mmulatta.UCSC.rheMac3.masked'\n67. 'BSgenome.Mmulatta.UCSC.rheMac8'\n68. 'BSgenome.Mmusculus.UCSC.mm10'\n69. 'BSgenome.Mmusculus.UCSC.mm10.masked'\n70. 'BSgenome.Mmusculus.UCSC.mm8'\n71. 'BSgenome.Mmusculus.UCSC.mm8.masked'\n72. 'BSgenome.Mmusculus.UCSC.mm9'\n73. 'BSgenome.Mmusculus.UCSC.mm9.masked'\n74. 'BSgenome.Osativa.MSU.MSU7'\n75. 'BSgenome.Ptroglodytes.UCSC.panTro2'\n76. 'BSgenome.Ptroglodytes.UCSC.panTro2.masked'\n77. 'BSgenome.Ptroglodytes.UCSC.panTro3'\n78. 'BSgenome.Ptroglodytes.UCSC.panTro3.masked'\n79. 'BSgenome.Ptroglodytes.UCSC.panTro5'\n80. 'BSgenome.Ptroglodytes.UCSC.panTro6'\n81. 'BSgenome.Rnorvegicus.UCSC.rn4'\n82. 'BSgenome.Rnorvegicus.UCSC.rn4.masked'\n83. 'BSgenome.Rnorvegicus.UCSC.rn5'\n84. 'BSgenome.Rnorvegicus.UCSC.rn5.masked'\n85. 'BSgenome.Rnorvegicus.UCSC.rn6'\n86. 'BSgenome.Scerevisiae.UCSC.sacCer1'\n87. 'BSgenome.Scerevisiae.UCSC.sacCer2'\n88. 'BSgenome.Scerevisiae.UCSC.sacCer3'\n89. 'BSgenome.Sscrofa.UCSC.susScr11'\n90. 'BSgenome.Sscrofa.UCSC.susScr3'\n91. 'BSgenome.Sscrofa.UCSC.susScr3.masked'\n92. 'BSgenome.Tgondii.ToxoDB.7.0'\n93. 'BSgenome.Tguttata.UCSC.taeGut1'\n94. 'BSgenome.Tguttata.UCSC.taeGut1.masked'\n95. 'BSgenome.Tguttata.UCSC.taeGut2'\n96. 'BSgenome.Vvinifera.URGI.IGGP12Xv0'\n97. 'BSgenome.Vvinifera.URGI.IGGP12Xv2'\n98. 'BSgenome.Vvinifera.URGI.IGGP8X'\n\n\n","text/html":["<style>\n",".list-inline {list-style: none; margin:0; padding: 0}\n",".list-inline>li {display: inline-block}\n",".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n","</style>\n","<ol class=list-inline><li>'BSgenome.Alyrata.JGI.v1'</li><li>'BSgenome.Amellifera.BeeBase.assembly4'</li><li>'BSgenome.Amellifera.UCSC.apiMel2'</li><li>'BSgenome.Amellifera.UCSC.apiMel2.masked'</li><li>'BSgenome.Aofficinalis.NCBI.V1'</li><li>'BSgenome.Athaliana.TAIR.04232008'</li><li>'BSgenome.Athaliana.TAIR.TAIR9'</li><li>'BSgenome.Btaurus.UCSC.bosTau3'</li><li>'BSgenome.Btaurus.UCSC.bosTau3.masked'</li><li>'BSgenome.Btaurus.UCSC.bosTau4'</li><li>'BSgenome.Btaurus.UCSC.bosTau4.masked'</li><li>'BSgenome.Btaurus.UCSC.bosTau6'</li><li>'BSgenome.Btaurus.UCSC.bosTau6.masked'</li><li>'BSgenome.Btaurus.UCSC.bosTau8'</li><li>'BSgenome.Btaurus.UCSC.bosTau9'</li><li>'BSgenome.Carietinum.NCBI.v1'</li><li>'BSgenome.Celegans.UCSC.ce10'</li><li>'BSgenome.Celegans.UCSC.ce11'</li><li>'BSgenome.Celegans.UCSC.ce2'</li><li>'BSgenome.Celegans.UCSC.ce6'</li><li>'BSgenome.Cfamiliaris.UCSC.canFam2'</li><li>'BSgenome.Cfamiliaris.UCSC.canFam2.masked'</li><li>'BSgenome.Cfamiliaris.UCSC.canFam3'</li><li>'BSgenome.Cfamiliaris.UCSC.canFam3.masked'</li><li>'BSgenome.Cjacchus.UCSC.calJac3'</li><li>'BSgenome.Dmelanogaster.UCSC.dm2'</li><li>'BSgenome.Dmelanogaster.UCSC.dm2.masked'</li><li>'BSgenome.Dmelanogaster.UCSC.dm3'</li><li>'BSgenome.Dmelanogaster.UCSC.dm3.masked'</li><li>'BSgenome.Dmelanogaster.UCSC.dm6'</li><li>'BSgenome.Drerio.UCSC.danRer10'</li><li>'BSgenome.Drerio.UCSC.danRer11'</li><li>'BSgenome.Drerio.UCSC.danRer5'</li><li>'BSgenome.Drerio.UCSC.danRer5.masked'</li><li>'BSgenome.Drerio.UCSC.danRer6'</li><li>'BSgenome.Drerio.UCSC.danRer6.masked'</li><li>'BSgenome.Drerio.UCSC.danRer7'</li><li>'BSgenome.Drerio.UCSC.danRer7.masked'</li><li>'BSgenome.Dvirilis.Ensembl.dvircaf1'</li><li>'BSgenome.Ecoli.NCBI.20080805'</li><li>'BSgenome.Gaculeatus.UCSC.gasAcu1'</li><li>'BSgenome.Gaculeatus.UCSC.gasAcu1.masked'</li><li>'BSgenome.Ggallus.UCSC.galGal3'</li><li>'BSgenome.Ggallus.UCSC.galGal3.masked'</li><li>'BSgenome.Ggallus.UCSC.galGal4'</li><li>'BSgenome.Ggallus.UCSC.galGal4.masked'</li><li>'BSgenome.Ggallus.UCSC.galGal5'</li><li>'BSgenome.Ggallus.UCSC.galGal6'</li><li>'BSgenome.Hsapiens.1000genomes.hs37d5'</li><li>'BSgenome.Hsapiens.NCBI.GRCh38'</li><li>'BSgenome.Hsapiens.UCSC.hg17'</li><li>'BSgenome.Hsapiens.UCSC.hg17.masked'</li><li>'BSgenome.Hsapiens.UCSC.hg18'</li><li>'BSgenome.Hsapiens.UCSC.hg18.masked'</li><li>'BSgenome.Hsapiens.UCSC.hg19'</li><li>'BSgenome.Hsapiens.UCSC.hg19.masked'</li><li>'BSgenome.Hsapiens.UCSC.hg38'</li><li>'BSgenome.Hsapiens.UCSC.hg38.masked'</li><li>'BSgenome.Mdomestica.UCSC.monDom5'</li><li>'BSgenome.Mfascicularis.NCBI.5.0'</li><li>'BSgenome.Mfuro.UCSC.musFur1'</li><li>'BSgenome.Mmulatta.UCSC.rheMac10'</li><li>'BSgenome.Mmulatta.UCSC.rheMac2'</li><li>'BSgenome.Mmulatta.UCSC.rheMac2.masked'</li><li>'BSgenome.Mmulatta.UCSC.rheMac3'</li><li>'BSgenome.Mmulatta.UCSC.rheMac3.masked'</li><li>'BSgenome.Mmulatta.UCSC.rheMac8'</li><li>'BSgenome.Mmusculus.UCSC.mm10'</li><li>'BSgenome.Mmusculus.UCSC.mm10.masked'</li><li>'BSgenome.Mmusculus.UCSC.mm8'</li><li>'BSgenome.Mmusculus.UCSC.mm8.masked'</li><li>'BSgenome.Mmusculus.UCSC.mm9'</li><li>'BSgenome.Mmusculus.UCSC.mm9.masked'</li><li>'BSgenome.Osativa.MSU.MSU7'</li><li>'BSgenome.Ptroglodytes.UCSC.panTro2'</li><li>'BSgenome.Ptroglodytes.UCSC.panTro2.masked'</li><li>'BSgenome.Ptroglodytes.UCSC.panTro3'</li><li>'BSgenome.Ptroglodytes.UCSC.panTro3.masked'</li><li>'BSgenome.Ptroglodytes.UCSC.panTro5'</li><li>'BSgenome.Ptroglodytes.UCSC.panTro6'</li><li>'BSgenome.Rnorvegicus.UCSC.rn4'</li><li>'BSgenome.Rnorvegicus.UCSC.rn4.masked'</li><li>'BSgenome.Rnorvegicus.UCSC.rn5'</li><li>'BSgenome.Rnorvegicus.UCSC.rn5.masked'</li><li>'BSgenome.Rnorvegicus.UCSC.rn6'</li><li>'BSgenome.Scerevisiae.UCSC.sacCer1'</li><li>'BSgenome.Scerevisiae.UCSC.sacCer2'</li><li>'BSgenome.Scerevisiae.UCSC.sacCer3'</li><li>'BSgenome.Sscrofa.UCSC.susScr11'</li><li>'BSgenome.Sscrofa.UCSC.susScr3'</li><li>'BSgenome.Sscrofa.UCSC.susScr3.masked'</li><li>'BSgenome.Tgondii.ToxoDB.7.0'</li><li>'BSgenome.Tguttata.UCSC.taeGut1'</li><li>'BSgenome.Tguttata.UCSC.taeGut1.masked'</li><li>'BSgenome.Tguttata.UCSC.taeGut2'</li><li>'BSgenome.Vvinifera.URGI.IGGP12Xv0'</li><li>'BSgenome.Vvinifera.URGI.IGGP12Xv2'</li><li>'BSgenome.Vvinifera.URGI.IGGP8X'</li></ol>\n"]},"metadata":{"tags":[]}}]},{"cell_type":"markdown","metadata":{"id":"SWf6NRI9yWKc"},"source":["## 2. Biostrings and when to use them? --zikaVirus as an example"]},{"cell_type":"markdown","metadata":{"id":"gA-9cA0FyeFJ"},"source":["### 2.1 Introduction to Biostrings"]},{"cell_type":"code","metadata":{"id":"gV6s9ueLyZU7"},"source":["# install Biostrings package\n","BiocManager::install(\"Biostrings\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"SMzDsAwG1orm"},"source":["#### 2.1.1 Exploring sequences\n","\n","zikaVirus fasta file could be download from NCBI [zikaVirus whole Genome](https://www.ncbi.nlm.nih.gov/nuccore/NC_012532.1?report=fasta)"]},{"cell_type":"code","metadata":{"id":"D8ynEw2uypdZ"},"source":["# Load packages\n","library(Biostrings)\n","\n","# read zikaVirus genome sequence that downloaded from NCBI\n","zikaVirus <- readDNAStringSet('/content/sequence.fasta')\n","\n","# check the alphabet of the zikaVirus\n","alphabet(zikaVirus)\n","\n","# check alphabetFrequency of the zikaVirus\n","alphabetFrequency(zikaVirus)\n","\n","# check alphabet using baseOnly = TRUE (only show base letter)\n","alphabet(zikaVirus, baseOnly = TRUE)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"thVuRdXu1tK7"},"source":["#### 2.1.2 Manipulating Biostrings"]},{"cell_type":"code","metadata":{"id":"5ckAstbR1xfe"},"source":["# Unlist the set, select the first 21 letters, and assign to dna_seq\n","dna_seq <- subseq(unlist(zikaVirus), end = 21)\n","dna_seq\n","\n","# Transcribe dna_seq into an RNAString object and print it\n","rna_seq <- RNAString(dna_seq) \n","rna_seq\n","\n","# Translate rna_seq into an AAString object and print it\n","aa_seq <- translate(rna_seq)\n","aa_seq\n","\n","\n","# shortcut translate DNA to amino acid\n","aa_seq <- translate(dna_seq)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"BmOzqmPd3f4Z"},"source":["### 2.2 Sequence Handling"]},{"cell_type":"markdown","metadata":{"id":"DKpIgk1-4XL_"},"source":["####2.2.1 From a single sequence to a set"]},{"cell_type":"code","metadata":{"id":"1jDYYaOp4bQ1"},"source":["# to create a new set from a single sequence\n","zikaSet <- DNAStringSet(zikaVirus, start = c(1, 101, 201), end = c(100, 200, 300))\n","zikaSet"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"owxLJN0oBz5I"},"source":["#### 2.2.2 From a set to a single sequence"]},{"cell_type":"code","metadata":{"id":"IlW3WTzF2-x_"},"source":["# From a set to a single sequence\n","zikv <- unlist(zikaSet)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"peO1b6U-EKGR"},"source":["#### 2.2.3 Common sequence manipulation functions"]},{"cell_type":"code","metadata":{"id":"9wJSI2p8ENk5"},"source":["# Reverse the zikv sequence\n","reverse(zikv)\n","\n","# Complement the zikv sequence\n","complement(zikv)\n","\n","# Reverse complement the zikv sequence\n","reverseComplement(zikv)\n","\n","# Translate the zikv sequence\n","translate(zikv)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"53I74UasE0rX"},"source":["#### 2.2.4 finding Pattern"]},{"cell_type":"code","metadata":{"id":"mFyktFkgE3Fy"},"source":["# 1. Find palindromes in zikv\n","findPalindromes(zikv)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"ryFo7v3CFGIe"},"source":["# 2. Find a conserved region within six frames\n","\n","# print the rnaframesZikaSet\n","rnaframesZikaSet\n","\n","# translate all 6 reading frames \n","AAzika6F <- translate(rnaframesZikaSet)\n","AAzika6F\n","\n","# Count the matches allowing 15 mistmatches\n","vcountPattern(pattern = NS5, subject = AAzika6F, max.mismatch = 15)\n","\n","# Select the frame that contains the match\n","selectedSet <- AAzika6F[3] \n","  \n","#Convert this frame into a single sequence\n","selectedSeq <- unlist(selectedSet)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"L6HN0FGSGKYR"},"source":["# 3. Looking for a match\n","\n","# Use vmatchPattern() with the set\n","vmatchPattern(pattern = ns5, subject = selectedSet, max.mismatch = 15)\n","\n","# Use matchPattern() with the single sequence\n","matchPattern(pattern = ns5, subject = selectedSeq, max.mismatch = 15)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"0Zvy6NyH7cdK"},"source":["## 3. IRanges and GenomicRanges"]},{"cell_type":"markdown","metadata":{"id":"MRdCgDNo8rdB"},"source":["###3.1 IRanges and Genomic Structures"]},{"cell_type":"markdown","metadata":{"id":"FKGPyKlWAS_M"},"source":["#### 3.1.1 Create an IRganges"]},{"cell_type":"code","metadata":{"id":"3Nf_UwV97f3H"},"source":["library(IRanges)\n","library(GenomicRanges)\n","\n","# start vector 1 through 5, end 100\n","IRnum1 <- IRanges(start = 1:5, end = 100)\n","\n","# end 100, width 89 and 10\n","IRnum2 <- IRanges(end = 100, width = c(89, 10))\n","\n","# logical argument start = Rle(c(F, T, T, T, F, T, T, T))\n","IRlog1 <- IRganges(start = Rle(c(F,T,T,T,F,T,T,T)))\n","\n","# Printing objects in a list\n","print(list(IRnum1 = IRnum1, IRnum2= IRnum2, IRlog1 = IRlog1))\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"WRsawSoWAXjm"},"source":["#### 3.1.2 Interacting with IRanges"]},{"cell_type":"code","metadata":{"id":"0z4yM2fNAbAU"},"source":["# Create the first sequence seq_1\n","seq_1 <- IRanges(start = 10, end = 37)\n","\n","# Create the second sequence seq_2\n","seq_2 <- IRanges(start = c(5, 35, 50),\n","                 end = c(12, 39, 61),\n","                 names = LETTERS[1:3])\n","\n","# Check the width of seq_1 and seq_2\n","width(seq_1)\n","width(seq_2)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"GFqozGAZBE_f"},"source":["### 3.2 Gene of Interest"]},{"cell_type":"markdown","metadata":{"id":"apZpUB4RBNxl"},"source":["#### 3.2.1 From tabular data to Genomic Ranges"]},{"cell_type":"code","metadata":{"id":"IXXsoZpzA2Jg"},"source":["# Load package GenomicRanges\n","library(GenomicRanges)\n","\n","# Print seq_intervals\n","seq_intervals\n","\n","# Create myGR\n","myGR <- as(seq_intervals, \"GRanges\")\n","\n","# Print myGR\n","myGR"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"xtVMU-IgCXki"},"source":["#### 3.2.2 GenomicRanges accessors"]},{"cell_type":"code","metadata":{"id":"MHRn1X9WCa94"},"source":["# Load GenomicRanges\n","library(GenomicRanges)\n","\n","# Print the seqinfo of object myGR\n","seqinfo(myGR)\n","\n","# Check the metadata, if any\n","mcols(myGR)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ZMBGpqF1DDbO"},"source":["### 3.3 Case study: ABCD1 mutation in human genome chromosome X"]},{"cell_type":"markdown","metadata":{"id":"Rlxn4ZYuDMhi"},"source":["#### 3.3.1 Manipulating collections of GRanges"]},{"cell_type":"code","metadata":{"id":"GeSsHawYDJ6m"},"source":["# Load human reference genome hg38\n","library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n","\n","# Assign hg38 to hg, then print it\n","hg <- TxDb.Hsapiens.UCSC.hg38.knownGene\n","hg"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ptp1PieWAAKM"},"source":["#### 3.3.2 overlapping of ABCD1 gene with chromosome"]},{"cell_type":"code","metadata":{"id":"AqT-jqfIlOR2"},"source":["# Store the overlapping range in rangefound\n","rangefound <- subsetByOverlaps(hg_chrX, ABCD1)\n","\n","# Check names of rangefound\n","names(rangefound)\n","\n","# Check the gene of interest \n","ABCD1\n","\n","# Check rangefound\n","rangefound"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"8C99Q60_KgTJ"},"source":["# Load the human transcripts DB to hg\n","library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n","hg <- TxDb.Hsapiens.UCSC.hg38.knownGene\n","\n","# Prefilter chromosome X \"chrX\" using seqlevels()\n","seqlevels(hg) <- c(\"chrX\")\n","\n","# Get all transcripts by gene and print it\n","hg_chrXt <- transcriptsBy(hg, by = \"gene\")\n","hg_chrXt\n","\n","# Select gene `215` from the transcripts\n","hg_chrXt$`215`"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"hi3AdQx1uO2n"},"source":["#### 3.3.3 From GRangesList object into a GRanges object"]},{"cell_type":"code","metadata":{"id":"kfdB1uhzuNkc"},"source":["# Unlist hg_ChrX and save result as myGR\n","myGR <- unlist(hg_ChrX)\n","\n","# Compare classes of hg_ChrX and myGR\n","class(hg_ChrX)\n","class(myGR)\n","\n","# Compare length of hg_ChrX and myGR\n","length(hg_ChrX)\n","length(myGR)\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"NCaoIT3Sxvzy"},"source":["## 4. Introducing ShortRead"]},{"cell_type":"markdown","metadata":{"id":"bedyRD2Ox6GS"},"source":["###4.1 Exploring and extract sequence files-fastq\n","1. Explore fastq file\n","2. Extract a sample from a fastq file"]},{"cell_type":"code","metadata":{"id":"JWLhaKpkx2Cv"},"source":["# Load ShortRead\n","library(ShortRead)\n","\n","# Print fqsample\n","fqsample\n","\n","# Check class of fqsample\n","class(fqsample)\n","\n","# Check class sread fqsample\n","class(sread(fqsample))\n","\n","# Check id of fqsample\n","id(fqsample)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"l9yxkGB70y64"},"source":["# Extract a sample from a fasts file (resampling)\n","# Load ShortRead\n","library(ShortRead)\n","\n","# Set a seed for sampling\n","set.seed(1234)\n","\n","# Use FastqSampler with f and select 100 reads\n","fs <- FastqSampler(con = f, n = 100)\n","\n","# Generate new sample yield\n","my_sample <- yield(fs)\n","\n","# Print my_sample\n","my_sample"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"A3CdpBLE1Nb5"},"source":["### 4.2 Sequence Quality\n","1. Base quality plot\n","2. nucleotide frequency plot (cycle vs count)"]},{"cell_type":"code","metadata":{"id":"Qs5fNZYM1NCk"},"source":["# Glimpse nucByCycle\n","glimpse(nucByCycle)\n","\n","# Create a line plot of cycle vs count\n","nucByCycle %>% \n","  # Gather the nucleotide letters in alphabet and get a new count column\n","  gather(key = alphabet, value = count , -cycle) %>% \n","  ggplot(aes(x = cycle, y =  count, color = alphabet)) +\n","  geom_line(size = 0.5 ) +\n","  labs(y = \"Frequency\") +\n","  theme_bw() +\n","  theme(panel.grid.major.x = element_blank())"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"jeUOq0HO72wW"},"source":["### 4.3 Match and filter\n","1. match and filter\n","2. Removing duplicates\n","3. more filtering"]},{"cell_type":"code","metadata":{"id":"yvcnZA4275iM"},"source":["# a filter those reads that start with the pattern \"ATGCA\"\n","myStartFilter <- srFilter(function(x) substr(sread(x), 1, 5) == \"ATGCA\")\n","\n","# Load package ShortRead\n","library(ShortRead)\n","\n","# Check class of fqsample\n","class(fqsample)\n","\n","# Filter reads into selectedReads using myStartFilter\n","selectedReads <- fqsample[myStartFilter(fqsample)]\n","\n","# Check class of selectedReads\n","class(selectedReads)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"ZGhD9Nay83qi"},"source":["# Sample with duplicates of class: ShortReadQ\n","dfqsample\n","\n","# Get the reads from dfqsample\n","mydReads <- sread(dfqsample)\n","\n","# Counting duplicates\n","table(srduplicated(mydReads))\n","\n","# filtering duplicated reads in a file\n","mydReads[srduplicated(mydReads) == FALSE]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"ixA-0dDdAUbG"},"source":["# more filter\n","\n","# Check reads of fqsample\n","sread(fqsample)\n","\n","# Create myFil using polynFilter\n","myFil <- polynFilter(threshold = 3, nuc = c(\"A\"))\n","\n","# Apply your filter to fqsample\n","filterCondition <- myFil(fqsample)\n","\n","# Use myFil with fqsample\n","filteredSequences <- fqsample[filterCondition]\n","\n","# Check reads of filteredSequences\n","sread(filteredSequences)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"dVvYhA9jELJZ"},"source":["### 4.4 Multiple assessment--library(Rqc)\n"]},{"cell_type":"code","metadata":{"id":"t-SmeiEOEPVP"},"source":["# Load package Rqc\n","library(Rqc)\n","\n","# Average per cycle quality plot\n","rqcCycleAverageQualityPlot(qa)\n","\n","# Average per cycle quality plot with white background\n","rqcCycleAverageQualityPlot(qa) + theme_minimal()\n","\n","# Read quality plot with white background\n","rqcReadQualityPlot(qa) + theme_minimal()\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"QFDo5FQvHNqo"},"source":[""],"execution_count":null,"outputs":[]}]}